lib/ts/History.h

```diff
@@ -0,0 +1,81 @@
+/** @file
+ *
+ *  A brief file description
+ *
+ *  @section license License
+ *
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef HISTORY_H_84E66E8E_1D2A_438B_A09A_2CD8FB51B209
+#define HISTORY_H_84E66E8E_1D2A_438B_A09A_2CD8FB51B209
+
+#include "ts/SourceLocation.h"
+
+#define NO_REENTRANT 99999
+#define HISTORY_DEFAULT_SIZE 65
+
+struct HistoryEntry {
+  SourceLocation location;
+  unsigned short event = 0;
+  short reentrancy     = 0;
+};
+
+template <unsigned Count> class History
+{
+public:
+  HistoryEntry history[Count];
+
+  void
+  push_back(const SourceLocation &location, int event, int reentrant = NO_REENTRANT)
+  {
+    int pos                 = history_pos++ % history_length;
+    history[pos].location   = location;
+    history[pos].event      = (unsigned short)event;
+    history[pos].reentrancy = (short)reentrant;
+  }
+
+  void
+  clear()
+  {
+    ink_zero(history);
+    history_pos = 0;
+  }
+
+  unsigned int
+  get_size() const
+  {
+    return history_pos > history_length ? history_length : history_pos;
+  }
+
+  const HistoryEntry &operator[](unsigned int i) { return history[i]; }
+
+  unsigned int history_pos          = 0;
+  const unsigned int history_length = Count;
+};
+
+#define REMEMBER(e, r)                             \
+  {                                                \
+    history.push_back(MakeSourceLocation(), e, r); \
+  }
+
+#define SM_REMEMBER(sm, e, r)                          \
+  {                                                    \
+    sm->history.push_back(MakeSourceLocation(), e, r); \
+  }
```

Comment:
I'd prefer these macros to be in the source files or headers that need them for 2 reasons
1. REMEMBER is a pretty generic name and price to collisions
2. They create a (logical) dependency from lists up to proxy, which we should avoid.